File: 백준/Gold/10026. 적록색약/적록색약.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

class Main {

    static int N;
    static char[][] grid;
    static boolean[][] visited;
    static int count;
    static int[] dx = {1, -1, 0, 0};
    static int[] dy = {0, 0, 1, -1};


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        N = Integer.parseInt(br.readLine());
        grid = new char[N][N];
        visited = new boolean[N][N];
        count = 0;

        for (int i = 0; i < N; i++) {
            String S = br.readLine();
            for (int j = 0; j < S.length(); j++) {
                grid[i][j] = S.charAt(j);
            }

        }
//        System.out.println(Arrays.deepToString(grid));

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (!visited[i][j]) {
                    count++;
                    dfs(i, j);
                }


            }

        }

        sb.append(count + " ");
        count = 0;
        visited = new boolean[N][N];

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (grid[i][j] == 'G') grid[i][j] = 'R';

            }

        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (!visited[i][j]) {
                    count++;
                    dfs(i, j);
                }

            }
        }

        sb.append(count);

        System.out.println(sb);


            }

    private static void dfs(int x, int y) {
        char c = grid[x][y];
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (0 > nx || nx >= N || 0 > ny || ny >= N || visited[nx][ny]) continue;
            if (c == grid[nx][ny]) {
                visited[nx][ny] = true;
                dfs(nx, ny);
            }

        }

    }
}

File: 백준/Gold/10026. 적록색약/적록색약.py
import sys

sys.setrecursionlimit(10**6)
input = sys.stdin.readline

N = int(input())

grid = [input().strip() for _ in range(N)]

# print(grid)

visited = [[False] * N for _ in range(N)]

dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]

normal_count = 0


def dfs(x, y):
    a = grid[x][y]
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]
        if 0 > nx or nx >= N or 0 > ny or ny >= N or visited[nx][ny]:
            continue
        if a == grid[nx][ny]:
            visited[nx][ny] = True
            dfs(nx, ny)


for i in range(N):
    for j in range(N):
        if not visited[i][j]:
            normal_count += 1
            dfs(i, j)


color_weakness_count = 0
visited = [[False] * N for _ in range(N)]

for i in range(N):
    grid[i] = grid[i].replace('G', 'R')

# print(grid)


for i in range(N):
    for j in range(N):
        if not visited[i][j]:
            color_weakness_count += 1
            dfs(i, j)


print(normal_count, color_weakness_count)